### ISSUE TITLE
Training on initial label file is very slow (compared to training on pkg)

### ISSUE THREAD

**Issue by olivier-cuttlefish** [talmolab/sleap#2431]
# Bug description
In SLEAP v1.5.1, training a model with the original label .slp file is very slow (0.5 it/s). Using the exact same config, but training on an exported .pkg.slp file, the speed is normal (~4 it/s). Interestingly, when running the same training from sleap v1.4.1, I do not run into the issue of slow training. I assume it must be something related to having a high number of videos in my label file and data loading. 

I run 
```
uv run sleap-nn-train --config-name centroid.yaml  --config-dir ~/.../sleap_files/training_251020_1453
```

I attached my config file. To compare, I just change the path from the .slp to .pkg.slp (see comments).
[centroid.yaml](https://github.com/user-attachments/files/22996224/centroid.yaml)

# System
Ubuntu 20.04.6 LTS
RTX 3060 12GB, Driver 560.35.05, Cuda: 12.6. 

# Installation method:
```uv python pin 3.12
mkdir sleap
cd sleap
uv venv --python 3.12
uv init
uv pip install "sleap[nn]" --index https://download.pytorch.org/whl/cu126 --index https://pypi.org/simple```

**Comment by gitttt-1234** [comment]
Hi @olivier-cuttlefish !

You could speed-up the training by setting `data_config.data_pipeline_fw: torch_dataset_cache_img_memory` and increase `num_workers` in the config to > 0 (maybe 2 or 4). This would preload the images into memory which will speed up the training.

Thanks,

Divya

**Comment by Heinrich2818** [comment]
> data_config.data_pipeline_fw: torch_dataset_cache_img_memory

Hi @gitttt-1234 !
I had really slow inference speed(2-3 fps, no tracker) in SLEAP v1.5.2. Is it possible to increase the inference speed by setting `data_config.data_pipeline_fw: torch_dataset_cache_img_memory `?

**Comment by talmo** [comment]
Hi @olivier-cuttlefish, thanks for the detailed report — the attached config made this easy to pin down.

**What's happening:** with `data_pipeline_fw: torch_dataset`, SLEAP-NN reads each frame straight from the video on every training step, and that loader runs single-threaded (`num_workers: 0`, since video backends can't be shared across worker processes).

- Raw **`.slp`** → every frame is a random-access seek + decode into your original compressed videos; with many videos that I/O dominates and the GPU starves → ~0.5 it/s.
- **`.pkg.slp`** → frames are embedded as an HDF5 image dataset → fast reads → ~4 it/s.
- SLEAP v1.4.1 used a TensorFlow `tf.data` pipeline that prefetched on background threads, which hid this cost.

So it's really "stream from source video" vs "pre-load frames", exactly as @gitttt-1234 noted. **Fix** — cache the images:

```yaml
data_config:
  data_pipeline_fw: torch_dataset_cache_img_memory   # or ..._cache_img_disk if it won't fit in RAM
trainer_config:
  train_data_loader: {num_workers: 4}   # >0 is fine once caching is on
  val_data_loader: {num_workers: 4}
```

This is also fixed by default: as of **SLEAP v1.6.0** the training dialog defaults the Data Pipeline Framework to **"Cache in Memory"**, so newer GUI-generated configs already use this. Upgrading to the latest (v1.6.3 — https://docs.sleap.ai/latest/installation) gives you the fast path automatically.

@Heinrich2818 — caching (`data_pipeline_fw`) is a *training* data-loader setting and won't affect inference speed. Slow inference is a separate matter (batch size, device, model/input size, video decode) — could you open a new issue with your `sleap-nn system` output and the command you ran? Happy to dig in there.

Closing as resolved — reopen if v1.6.3 still trains slowly for you.
